feat: SubscriptionManager to consolidate subscription handling

davidzhao opened the pull request:
Added a new manager to handle all subscription needs. Implemented using reconciler pattern. The goals are:
* improve subscription resilience by separating desired state and current state
* reduce complexity of synchronous processing
* better detect failures with the ability to trigger full reconnect

More comments inline.

davidzhao commented on pkg/rtc/participant.go:
```diff
@@ -92,6 +78,7 @@ type ParticipantParams struct {
 	ReconnectOnPublicationError  bool
 	ReconnectOnSubscriptionError bool
 	VersionGenerator             utils2.TimedVersionGenerator
+	TrackResolver                types.MediaTrackResolver
```
creating an interface for this to delegate responsibility of finding the track and permission checking to Room

davidzhao commented on pkg/rtc/participant.go:
```diff
@@ -460,12 +426,6 @@ func (p *ParticipantImpl) OnDataPacket(callback func(types.LocalParticipant, *li
 	p.lock.Unlock()
 }
 
-func (p *ParticipantImpl) OnSubscribedTo(callback func(types.LocalParticipant, livekit.ParticipantID)) {
```
moved to SubscriptionManager

davidzhao commented on pkg/rtc/participant.go:
```diff
@@ -684,20 +639,16 @@ func (p *ParticipantImpl) Close(sendLeave bool, reason types.ParticipantCloseRea
-
-		p.TransportManager.Close()
-	}()
+	go p.TransportManager.Close()
```
participant would only be concerned with Transport, downtracks would be closed by SubscriptionManager

davidzhao commented on pkg/rtc/participant.go:
```diff
@@ -872,25 +813,6 @@ func (p *ParticipantImpl) GetConnectionQuality() *livekit.ConnectionQualityInfo
 	}
 }
 
-func (p *ParticipantImpl) GetSubscribedParticipants() []livekit.ParticipantID {
```
moved to SubscriptionManager

davidzhao commented on pkg/rtc/participant.go:
```diff
@@ -963,133 +868,24 @@ func (p *ParticipantImpl) VerifySubscribeParticipantInfo(pID livekit.Participant
 }
 
-// AddSubscribedTrack adds a track to the participant's subscribed list
-func (p *ParticipantImpl) AddSubscribedTrack(subTrack types.SubscribedTrack, sourceTrack types.MediaTrack) {
```
reduced responsibilities of this function, and changed to onTrackSubscribed as a callback

davidzhao commented on pkg/rtc/participant.go:
```diff
@@ -963,133 +868,24 @@ func (p *ParticipantImpl) VerifySubscribeParticipantInfo(pID livekit.Participant
-		// have to send speaker update indicating that the participant speaker is no long active
-		// so that clients can clean up their speaker state for the leaving/unsubscribed participant
-		//
-		if p.ProtocolVersion().SupportsSpeakerChanged() {
```
this has been moved to Room, since it already handles sending updates when we subscribe to someone new.

davidzhao commented on pkg/rtc/room.go:
```diff
@@ -495,31 +507,40 @@ func (r *Room) RemoveDisallowedSubscriptions(sub types.LocalParticipant, disallo
-		pub.RemoveSubscriber(sub, trackID, false)
+		track := pub.GetPublishedTrack(trackID)
+		if track == nil {
+			track.RemoveSubscriber(sub.ID(), false)
```
when disallowed, immediately remove from track, reconciler will retry when permissions are granted again

boks1971 commented on pkg/rtc/room.go:
```diff
@@ -445,36 +457,36 @@ func (r *Room) UpdateSubscriptions(
-			if _, err := publisher.AddSubscriber(participant, types.AddSubscriberParams{TrackIDs: []livekit.TrackID{trackID}}); err != nil {
-				return err
-			}
+			participant.SubscribeToTrack(trackID, publisherIdentities[trackID], publisherID)
```
Can't say with certainty as it is hard to keep track in my head :-), but `SubscribeToTrack` may not be handling `nil` identity? It can be `nil` here. Is that handled properly? Or maybe this place should not call `SubscribeToTrack` if `identity` is `nil`?

davidzhao replied:
yeah it's possible for participantIdentity to be empty. We do not rely on publisher identity in the subscription stack, so this is just for improved logging.

boks1971 commented on pkg/rtc/participant.go:
```diff
@@ -755,16 +718,7 @@ func (p *ParticipantImpl) MaybeStartMigration(force bool, onStart func()) bool {
-		for _, dt := range downTracksToClose {
-			dt.CloseWithFlush(false)
-		}
+		p.SubscriptionManager.Close(true)
```
This is going to cause problems I think on migration. This one is closing tracks synchronously. That is required before closing the subscriber peer connection because of the reason mentioned in the comments ☝️ . But, subscription manager is closing tracks in a `goroutine`.

boks1971 reviewed (approved):
Great work @davidzhao !!!

Left a few small comments (as far as I could understand and more importantly follow all the changes :-) ). One case of synchronous operation vs goroutine that might cause trouble.

A couple of more things
- There was quite a bit of trickiness switching publishing source between `local` and `remote` media track. Gut feel: that is probably still going to throw some curve ball in this scheme and may require some tweaks.
- Subscription supervisor is now not required. Suggest deleting that code so that there is no confusion surrounding that.

davidzhao commented:
Thanks, these are great points. I'll take another pass at addressing some of these. Will ping you when it's ready.

davidzhao commented on pkg/rtc/subscriptionmanager_test.go:
```diff
@@ -0,0 +1,386 @@
+		}, subSettleTimeout, subCheckInterval, "track was not resubscribed")
+
+		// was subscribed twice, unsubscribed once (due to close)
+		require.Equal(t, int32(2), numParticipantSubscribed.Load())
```
this line will currently trigger the subscription to close

```
setTestSubscribedTrackClosed(t, s.getSubscribedTrack(), false)
```

for the participants subscribed to, we could either go with desired vs actual state. I went with the latter. So in this test, the subscribed track closure caused the subscription to be unsubscribed. That made it so that we do not perceive the participant to be actually subscribed to the publisher any longer.

however, reconciler would repair that condition and recreated the subscription.

boks1971 replied:
I kinda guessed that's what is happening 😄 . But, how could a subscription succeed again through the reconciler path if the published track is closed. Shouldn't the resolver return an empty track? That's what confused me, i. e. how did the reconciler make a successful subscription on a closed track?

davidzhao replied:
The published track wasn't closed, it was only the subscribedTrack. It's just a simulated test case to ensure reconciler is resilient to recreate the subscription.
In practice, the mediaTrack could have been closed, and this path will keep waiting until resolver could resolve the track again.

boks1971 replied:
oh got it. Thank you

davidzhao commented on pkg/rtc/subscriptionmanager_test.go:
```diff
@@ -0,0 +1,386 @@
+		// trackSubscribed telemetry not sent
+		tm := sm.params.Telemetry.(*telemetryfakes.FakeTelemetryService)
+		require.Equal(t, 1, tm.TrackSubscribeRequestedCallCount())
+		require.Equal(t, 0, tm.TrackSubscribedCallCount())
```
this is a good question. I think we should still treat permissions failures as failed events.. maybe logging them explicitly so that the user can be aware?

what do you think?

boks1971 replied:
Yes and no. 

yes for logging them clearly and give users the visibility.

no for treating them as failure. Saying no because system is behaving as expected, i. e. some subscription was desired, but there was no permission. For example spatial applications could enable auto-subscribe (intentionally or by mistake) and that could potentially load to lot of subscription attempts with a majority of them failing due to permissions as users are spread out in the space. My main concern is that attempts vs successes metric getting used to define health of a node. We can still count attempts. Maybe introduce a new counter like `permissionsFailed` and then do the failure percentage as `(numFailures - permissionsFailed) / numAttempts` and use that to effect any changes to node health.

wdyt? cc: @real-danm

davidzhao replied:
good question regarding what they will be used for. Perhaps the subscription itself can now generate different events:
* requested
* success
* no permission / track not found / etc -- failures that are outside of the responsibility of the subscription system
* subscription failure

Then we can choose to use these stats in different ways.

boks1971 replied:
Yeah, that would be my preference.

davidzhao commented:
@boks1971 added [this commit](https://github.com/livekit/livekit/pull/1317/commits/956cd527e7251010b1f5e2ea320d9bb6625b88f2) that addresses remaining issues, including changes to telemetry. CC: @real-danm

boks1971 commented on pkg/rtc/changenotifier.go:
```diff
@@ -0,0 +1,110 @@
+/*
```
Can this be move to `pkg/utils` or some other common place please? Nothing specific to `rtc` here and potentially can be used by others if necessary.

boks1971 commented on pkg/rtc/supervisor/participant_supervisor.go:
```diff
@@ -139,55 +122,6 @@ func (p *ParticipantSupervisor) ClearPublishedTrack(trackID livekit.TrackID, pub
 	p.lock.RUnlock()
 }
 
-func (p *ParticipantSupervisor) UpdateSubscription(trackID livekit.TrackID, isSubscribe bool, sourceTrack types.MediaTrack) {
```
Thank you @davidzhao for cleaning this up. There is a `OperationMonitorEvent` type in `interfaces.go` with events specific to subscription monitoring. Can you please delete the unused ones there also?

davidzhao replied:
got it, will do.

boks1971 commented on pkg/telemetry/events.go:
```diff
@@ -246,6 +246,8 @@ func (t *telemetryService) TrackSubscribeRequested(
 	publisher *livekit.ParticipantInfo,
 ) {
 	t.enqueue(func() {
+		prometheus.RecordTrackSubscribeAttempt()
```
Should we be recording the track type in the attempt also? Seeing it recorded for `success` below.

davidzhao replied:
due to subscription lifecycle changes, we do not have the track's type during the request. 

so I've removed kind from `promTrackSubscribeCounter`, since we are not using that attribute anywhere. but we would still log it for the current stat, so that we can graph what's currently being subscribed to.

boks1971 commented on pkg/telemetry/events.go:
```diff
@@ -269,12 +271,38 @@ func (t *telemetryService) TrackSubscribed(
+	t.enqueue(func() {
+		prometheus.RecordTrackUnsubscribed(track.Type.String())
+
+		if shouldSendEvent {
```
Is this not going to be sent in migration cases? Do we want one sent for migration also to get the full picture?

davidzhao replied:
I think ideally we would not show unsubscribe or subscribe events in the case of migrations. I think as it stands it'll still log Subscriptions again.. and that seems ok.

boks1971 replied:
From a reconciler perspective, I am guessing we want to have the switch from `local` -> `remote` or vice-versa as an attempt + success/failure so that it can take corrective action if there is a problem during the switch.

Whether we translate that to telemetry is something to think about. Yeah, not showing migration in telemetry sounds reasonable. But, wondering if we should have additional events to signify switch of source track from `local` -> `remote` and vice-versa.

davidzhao replied:
yeah this would be something to think about.. though this PR is getting a bit large. let's think about adding this insight in a future one.

boks1971 commented on pkg/telemetry/prometheus/rooms.go:
```diff
@@ -10,22 +10,25 @@ import (
-	participantTotal       atomic.Int32
-	trackPublishedTotal    atomic.Int32
-	trackSubscribedTotal   atomic.Int32
+	roomCurrent            atomic.Int32
```
why not change `promRoomTotal` -> `proomRoomCurrent` as that is also decrementing on close?

boks1971 commented on pkg/rtc/subscriptionmanager.go:
```diff
@@ -216,9 +215,29 @@ func (m *SubscriptionManager) OnSubscribeStatusChanged(fn func(publisherID livek
+		m.lock.RLock()
+		for _, sub := range m.subscriptions {
+			if sub.needsSubscribe() {
+				allSubscribed = false
```
guess you can break here as a single false means you have to go around.

boks1971 reviewed (commented):
A few more comments. LGTM! Great work!!!

boks1971 commented on pkg/utils/changenotifier.go:
```diff
@@ -21,49 +21,38 @@ import (
-
-	return len(n.observers) > 0
+	hasObservers := false
+	n.observers.Range(func(key, val any) bool {
```
@davidzhao What race caused this change? We have been removing use of `sync.Map` whenever there is a `Delete` involved due to [this](https://stackoverflow.com/questions/54174618/is-golang-sync-map-range-function-thread-safe).

davidzhao replied:
the race was in the notify code, where we have to call all of the callbacks. Thank you for the reminder about avoid sync.Map usage. I'll revert + fix it a diff way.
